### DISCUSSION TITLE
Is Nginx neccessary as the reverse proxy in front of my microservice app?

### DISCUSSION THREAD

**Discussion by u2ros** [BretFisher/ama#172] [Q&A]
Hi Bret,

In your courses about Node and Docker there's a lesson where Nginx is one of the reverse proxies you used in front of the sample microservice application. I've seen this recommended by others too, the reason being, so that you separate serving static content and API endpoints and improve performance. My question is. Do i really need to use Nginx (or any other reverse proxy) because in my project i have a requirement, where certain static content is only allowed to authorised users. In fact, the logic for which resource is accessible by which user is quite detailed and was so far done directly in Node.js (via passport, and some function checks). If i tried to replicate this behavior in Nginx it would be super complicated if not impossible. So, would using node.js as the application entry point container be considered a bad practice?

Thank you

**Comment by BretFisher** [comment]
To directly answer your question, no, it's not a bad practice to run a Node app directly on a published port. It just depends on your requirements.

I think you scenerio speaks to two slightly unrelated technologies:

1. A HTTP proxy, and a HTTP content cache. Nginx can do either or both.

2. A HTTP proxy (without caching) shouldn't change the behavior of your HTTP packets, and how connections are authenticated or served by your Node app.  A HTTP connection comes into the proxy, it inspects the header, and makes a decision on which backend container should receive that connection based on load balancing rules, domain name, url paths, etc. Common proxies for containers include NGINX, HAProxy, and the newer Traefik, which is "docker aware" and can auto-update its HTTP routing config based on the settings of other containers. These tools have been around since the dawn of web servers as optional solutions to various problems.

3. Your concerns sound like your thinking about content caching, which usually isn't enabled by default in HTTP proxies, and isn't even a feature of Traefik.  Web app typically don't need code changes to work with a proxy in front of it.  Technically, a simple AWS ALB is a proxy that accepts HTTP/S, and then retransmits the packet to the backend server. Think of a basic HTTP proxy as just a firewall and load balancer in one. It changes the "Client IP" and maybe adds a few proxy HTTP headers for convenience, maybe adds a cookie for session "stickyness", but doesn't relate to the way your HTTP app functions overall.

That all being said. NGINX, HTTP proxies, nor caching proxies are required for using Docker.  The reason you likely are hearing more about them is the fact that Docker/Swarm/K8s nodes allow you to run many containers on the same machine, rather than the usual "one web app per server". The nature of TCP/IP is that only one program can listen on a specific TCP port, so in the case of a Docker machine, that would restrict you to only one container working on 80/443. Having something like Nginx/Traefik use those ports, and then route connections to various containers on that server, means you can host many web containers, all which look like they are avaiaible on the server IP at ports 80/443, but in fact it's the proxy that's inspecting the header and changing the destination.
